### PR TITLE
CMake: Define BOOST_NO_CXX11_HDR_UNORDERED_MAP for bundled boost if <gcc-4.8

### LINKED ISSUE #3624: Incompatibility with 'emplace_hint' function and gcc versions below 4.8
This follows an issue raised on the ASPECT (https://github.com/geodynamics/aspect/issues/1271) GitHub page relating to deal.ii and the 'emplace_hint' function, which @bangerth was able to pinpoint. 

GCC versions prior to 4.8 do not contain the 'emplace_hint' function in map.h, but the checks in `boost/config/stdlib/libstdcpp3.hp` do not pick this up:
`//  C++0x headers in GCC 4.3.0 and later`
`//`
`#if (BOOST_LIBSTDCXX_VERSION < 40300) || !defined(BOOST_LIBSTDCXX11)`
`#  define BOOST_NO_CXX11_HDR_ARRAY`
`#  define BOOST_NO_CXX11_HDR_TUPLE`
`#  define BOOST_NO_CXX11_HDR_UNORDERED_MAP`
`#  define BOOST_NO_CXX11_HDR_UNORDERED_SET`
`#  define BOOST_NO_CXX11_HDR_FUNCTIONAL`
`#endif`

While deal.II compiles, to successfully compile the current developer version of ASPECT (uses of `emplace_hint` function) requires passing the following flag before calling `cmake` for deal.II:
`export CXXFLAGS=-DBOOST_NO_CXX11_HDR_UNORDERED_MAP`.

I'm not sure what the fix is here, but the issue has now been raised ;)

### PR DESCRIPTION
**Pull request by tamiko**
Closes #3624.

### CODE REVIEW

**Comment by tjhei** [comment]
Does this work if a code based on deal.II includes the offending boost header?

**Comment by tamiko** [comment]
@tjhei Yes [1]. That happens due to `LIST(APPEND BOOST_USER_DEFINITIONS "BOOST_ALL_NO_LIB")`.

Currently, the patch only sets the definition in case of *bundled* boost.


[1] If the code respects our recommendation of preprocessor definitions that gets exported in `DEAL_II_DEFINITIONS`.

**Comment by tjhei** [comment]
great.

> Currently, the patch only sets the definition in case of bundled boost.

What is the reasoning for that? I would assume that a standalone boost has the same problem.

**Comment by tamiko** [comment]
@tjhei Because you were worried about our preprocessor definition being too intrusive. The idea is that we do not work around the bug in external boost but rely on the distribution/user to fix the header.

I would have simply exported the definition in case of external boost as well.

**Comment by tjhei** [comment]
> @tjhei Because you were worried about our preprocessor definition being too intrusive.

Sorry, not liking the idea doesn't mean I am opposed. I was worried because we might break other stuff by making this define.

> I would have simply exported the definition in case of external boost as well.

Also fine with me. I let you decide.

**Comment by tamiko** [comment]
I have updated this PR. @bangerth Can you please check that the version ranges are correct?

File: cmake/configure/configure_boost.cmake [hunk 1]
```diff
@@ -47,6 +47,19 @@ MACRO(FEATURE_BOOST_CONFIGURE_BUNDLED)
     LIST(APPEND BOOST_DEFINITIONS "BOOST_ALL_NO_LIB")
     LIST(APPEND BOOST_USER_DEFINITIONS "BOOST_ALL_NO_LIB")
   ENDIF()
+
+  #
+  # Our bundled boost version 1.62 checks for the availability of
+  # "emplace_hint" incorrectly. Thus, simply define
```
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: -> incorrectly: It tests for the preprocessor define BOOST_NO_CXX11_HDR_UNORDERED_MAP in bundled/boost-.../include/boost/serialization/map.h thinking that that define is characteristic for the presence of std::(multi)map::emplace_hint. This is generally correct, except for GCC 4.8, for which the preprocessor variable is defined, but the function does not exist. (See also https://svn.boost.org/trac/boost/ticket/12755 .)

File: cmake/configure/configure_boost.cmake [hunk 2]
```diff
@@ -47,6 +47,19 @@ MACRO(FEATURE_BOOST_CONFIGURE_BUNDLED)
...
+  # BOOST_NO_CXX11_HDR_UNORDERED_MAP if the gcc compiler version is less
+  # than 4.8.
+  #
+  IF( CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND
+      CMAKE_CXX_COMPILER_VERSION VERSION_LESS "4.8" )
```
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: I think this is wrong -- 4.8 is the compiler that's affected. Is there a `VERSION_LESS_EQUAL`?

File: cmake/configure/configure_boost.cmake [hunk 3]
```diff
@@ -63,6 +76,17 @@ MACRO(FEATURE_BOOST_FIND_EXTERNAL var)
...
+    #
+    # Newer boost versions check for the availability of "emplace_hint"
+    # incorrectly. Thus, simply define BOOST_NO_CXX11_HDR_UNORDERED_MAP if
+    # the gcc compiler version is less than 4.8.
+    #
+    IF( CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_LESS "4.8" )
+      LIST(APPEND BOOST_DEFINITIONS "BOOST_NO_CXX11_HDR_UNORDERED_MAP")
+      LIST(APPEND BOOST_USER_DEFINITIONS "BOOST_NO_CXX11_HDR_UNORDERED_MAP")
+    ENDIF()
```
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: Could we externalize this into a macro so we don't have to duplicate the code?

File: cmake/configure/configure_boost.cmake (hunk 2 above)
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: I correct -- the problem is only up to 4.7.

File: cmake/configure/configure_boost.cmake (hunk 1 above)
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: ...except for GCC 4.7...

File: cmake/configure/configure_boost.cmake (hunk 3 above)
Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: For 5 lines? I would prefer to not define a macro for that.

File: cmake/configure/configure_boost.cmake (hunk 2 above)
Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: Thanks for verifying - so less than 4.8 was correct.

File: cmake/configure/configure_boost.cmake (hunk 3 above)
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: Alright, your call :-)

File: cmake/configure/configure_boost.cmake (hunk 2 above)
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: Yes. I went through the original bug reports again. It seems to work with 4.8, but not with 4.7.

**Comment by tamiko** [comment]
@bangerth Added comment and refactored the macros.

File: cmake/configure/configure_boost.cmake [hunk 4]
```diff
@@ -22,6 +22,31 @@ SET(DEAL_II_WITH_BOOST ON # Always true. We need it :-]
   )
 
 
+MACRO(FEATURE_BOOST_CONFIGURE_COMMON)
```
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: Nice ;-)

**Comment by bangerth** [comment]
Great, thanks!

OK to merge once tester is happy.

**Pull request merged**